packages/core/src/synthesis/Mutation.ts

```diff
@@ -0,0 +1,427 @@
+  additionalMutations: Mutation[];
+}
+
+export interface ChangeExprType extends IMutation {
+  tag: "ChangeExprType";
+  stmt: Bind;
+  expr: ArgExpr;
+  newStmt: SubStmt;
+  additionalMutations: Mutation[];
+}
+
+export const showOps = (ops: Mutation[]): string => {
+  return ops.map((op) => showOp(op)).join("\n");
+};
+
+export const showOp = (op: Mutation): string => {
+  switch (op.tag) {
+    // case "Replace":
+    //   return `Replace ${prettyStmt(op.old)} by ${prettyStmt(op.new)}`;
+    case "SwapStmtArgs":
+      return `Swap arguments of ${prettyStmt(op.stmt)}`;
+    case "SwapExprArgs":
+      return `Swap arguments of ${prettySubNode(op.expr)} in ${prettyStmt(
+        op.stmt
+      )}`;
+    case "ChangeStmtType":
+    case "ChangeExprType":
+      return `Change ${prettyStmt(op.stmt)} to ${prettyStmt(op.newStmt)}`;
+    case "ReplaceExprName":
+    case "ReplaceStmtName":
+      return `Replace the name of ${prettyStmt(op.stmt)} with ${op.newName}`;
+    // default:
+    case "Add":
+    case "Delete":
+      return `${op.tag} ${prettySubNode(op.stmt)}`;
+  }
+};
+
+//#endregion
+
+//#region Mutation execution
+
+export const executeMutation = (
+  mutation: Mutation,
+  prog: SubProg,
+  ctx: SynthesisContext
+): WithContext<SubProg> => mutation.mutate(mutation as any, prog, ctx); // TODO: typecheck this?
+
+export const executeMutations = (
+  mutations: Mutation[],
+  prog: SubProg,
+  ctx: SynthesisContext
+): WithContext<SubProg> =>
+  mutations.reduce(
+    ({ res, ctx }: WithContext<SubProg>, m: Mutation) =>
+      m.mutate(m as any, res, ctx),
+    { res: prog, ctx }
+  );
+
+const swap = (arr: any[], a: number, b: number) =>
+  arr.map((current, idx) => {
+    if (idx === a) return arr[b];
+    if (idx === b) return arr[a];
+    return current;
+  });
+
+//#endregion
+
+//#region Mutation guard functions
+
+const withCtx = <T>(res: T, ctx: SynthesisContext): WithContext<T> => ({
+  res,
+  ctx,
+});
+
+export const appendStmtCtx = (
+  { stmt }: Add,
+  p: SubProg,
+  ctx: SynthesisContext
+): WithContext<SubProg> => {
+  if (stmt.tag === "Decl") {
+    const newCtx = addID(ctx, stmt.type.name.value, stmt.name);
+    return withCtx(appendStmt(p, stmt), newCtx);
+  } else {
+    return withCtx(appendStmt(p, stmt), ctx);
+  }
+};
+
+export const removeStmtCtx = (
+  { stmt }: Delete,
+  prog: SubProg,
+  ctx: SynthesisContext
+): WithContext<SubProg> => {
+  if (stmt.tag === "Decl") {
+    const newCtx = removeID(ctx, stmt.type.name.value, stmt.name);
+    return withCtx(removeStmt(prog, stmt), newCtx);
+  } else {
+    return withCtx(removeStmt(prog, stmt), ctx);
+  }
+};
+
+export const checkAddStmts = (
+  prog: SubProg,
+  cxt: SynthesisContext,
+  newStmts: (cxt: SynthesisContext) => SubStmt[]
+): Add[] | undefined => {
+  const stmts: SubStmt[] = newStmts(cxt);
+  return stmts.map((stmt: SubStmt) => ({
+    tag: "Add",
+    stmt,
+    mutate: appendStmtCtx,
+  }));
+};
+
+export const checkAddStmt = (
```

Comment:
I'm only seeing this function being used in one place, it seems like it could just be entirely replaced by `checkAddStmts`

Reply:
Leaving it in, just in case future synthesizers will use this guard function.